keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,252 @@
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Request a progress notify from etcd on the etcd watch that keeps the cache up-to-date. This is needed because the watch only receives updates for the range of the keyspace being watched. So updates for KVs outside of that range are not observed by the watch, and so the highest revision observed by the watch can be lower than the current revision even if the watch is up-to-date. A successful progress notify solve this by sending an event that will the highest revision observed to the >= the current revision. The cache is then guaranteed to to sufficiently up-to-date to serve a request at the current revision.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough, we may not need to request progress notify events from etcd.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
+should the kube-apiserver also wait 20ms to a desired revision to become
+available or should it fallback to making the request to etcd sooner?
+
+Optional: For some or all of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+### Risks and Mitigations
+
+This can increase the number of round trips to etcd: One is required to get the
+latest revision, another might be needed to request the progress notify. Both of
+these requests are cheap and scalable, but we need to understand the latency
+impact of waiting for both of them before serving a response back to the
+client. If this turns out to have performance implications, it could be
+partially mitigated by configuring etcd to send regular progress events, which
+is something we will explore.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
```

Comment:
Should we provide an option like "RV=-1" which can skip "wait until fresh" and serving from etcd directly?

Because I think "wait until fresh" is not real "consistent read", as the data has been changed during the wait time.

Reply:
"Consistent read" here is similar to etcd quorum read, which in fact is exactly refers to linearizability consistency in distributed systems.

Reply:
+1, the consistency we're guaranteeing by waiting is exactly equivalent to the guarantee provided by linearizability consistency, so we can safely call it a consistent read. Note that in a distributed system there are many things that can result in a delay: network latency, gc pauses, etc.. Consistency isn't defined by how fast a response is returned, but by what properties it has.

Reply:
+1